### PR TITLE
Implemented load based task count estimator

### PR DESCRIPTION
**Pull request by jzakaryan**
Implementation of LoadBasedTaskCountEstimator and its tests.

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 1]
```diff
@@ -27,6 +41,22 @@
...
+
+    Set<String> allPartitions = new HashSet<>(assignedPartitions);
+    allPartitions.addAll(unassignedPartitions);
+
+    // total throughput in KB/sec
+    int totalThroughput = throughputMap.entrySet().stream().
+        filter(e -> allPartitions.contains(e.getKey())).mapToInt(e -> e.getValue().getBytesInKBRate()).sum();
+    LOG.debug("Total throughput in all partitions: {}KB/sec", totalThroughput);
+
+    int taskCountEstimate = (int) Math.ceil((double) totalThroughput / (TASK_CAPACITY_MBPS_DEFAULT * 1024));
```
Comment (shrinandthakkar) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:58]: do we know if the KBps rate we get from the provider is KB or KiB as there are a couple of conventions going around with that and based on it, we might have to change 1024 to 1000?

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 2]
```diff
@@ -27,6 +41,22 @@
...
+    Validate.notNull(assignedPartitions, "null assignedPartitions");
+    Validate.notNull(unassignedPartitions, "null unassignedPartitions");
+
+    Map<String, PartitionThroughputInfo> throughputMap = throughputInfo.getPartitionInfoMap();
+
+    Set<String> allPartitions = new HashSet<>(assignedPartitions);
+    allPartitions.addAll(unassignedPartitions);
+
+    // total throughput in KB/sec
+    int totalThroughput = throughputMap.entrySet().stream().
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:54]: 1. Should you account some default weight for the partitions for which you did not get entry in throughput map? For eg. We may setup query to get all the partitions with more than 25KBps, in which case, we can assume the partitions with <25KBps to have some default weight rather than 0.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 1 above)
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:58]: Along with `ceil`, I feel there is a need for some additional buffer. This task estimator should keep some scope for growth rather than tightly packing all the partitions in the task.

Since the kafka limit is 4MB, I think one way to get additional tasks is by either considering the default to be 3MB, which in other words means 75% throughput utilization and additional 25% buffer to grow, without increasing the task itself. 

The distribution will have some gaps/holes and it demands for some approximation.

The task estimator will need some smartness in these aspects

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 3]
```diff
@@ -27,6 +41,22 @@
    */
   public int getTaskCount(ClusterThroughputInfo throughputInfo, List<String> assignedPartitions,
       List<String> unassignedPartitions) {
-    throw new NotImplementedException();
+    Validate.notNull(throughputInfo, "null throughputInfo");
+    Validate.notNull(assignedPartitions, "null assignedPartitions");
+    Validate.notNull(unassignedPartitions, "null unassignedPartitions");
+
+    Map<String, PartitionThroughputInfo> throughputMap = throughputInfo.getPartitionInfoMap();
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:48]: What happens in the throughput info is not available ? Will it be the callers responsibility in that case? It will be better to add those information.

File: datastream-server/src/test/resources/partitionThroughput.json [hunk 4]
```diff
@@ -213,6 +213,16 @@
       "PeanutButter-7" : "bytesInKB: 8, msgIn:24",
       "PeanutButter-8" : "bytesInKB: 8, msgIn:26",
       "PeanutButter-9" : "bytesInKB: 8, msgIn:26"
+    },
+    "ice-cream" : {
+      "Vanilla-1" : "bytesInKB: 1024, msgIn:100",
+      "Vanilla-2" : "bytesInKB: 1024, msgIn:100",
+      "Vanilla-3" : "bytesInKB: 1024, msgIn:100",
+      "Vanilla-4" : "bytesInKB: 1024, msgIn:100",
```
Comment (vmaheshw) [inline datastream-server/src/test/resources/partitionThroughput.json:221]: pick a case, where the byteInKB is more than the max supported in one task.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 5]
```diff
@@ -5,18 +5,32 @@
...
 /**
  * Estimates the minimum number of tasks for a datastream based on per-partition throughput information
  */
 public class LoadBasedTaskCountEstimator {
 
+  private static final Logger LOG = LoggerFactory.getLogger(LoadBasedTaskCountEstimator.class.getName());
+
+  // TODO Make this constant configurable
+  @VisibleForTesting
+  public final static int TASK_CAPACITY_MBPS_DEFAULT = 4;
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:32]: since it is visible for testing should you remove the `final` ?

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 6]
```diff
@@ -27,6 +41,22 @@
...
+
+    Map<String, PartitionThroughputInfo> throughputMap = throughputInfo.getPartitionInfoMap();
+
+    Set<String> allPartitions = new HashSet<>(assignedPartitions);
+    allPartitions.addAll(unassignedPartitions);
+
+    // total throughput in KB/sec
+    int totalThroughput = throughputMap.entrySet().stream().
+        filter(e -> allPartitions.contains(e.getKey())).mapToInt(e -> e.getValue().getBytesInKBRate()).sum();
+    LOG.debug("Total throughput in all partitions: {}KB/sec", totalThroughput);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:56]: Set this level to INFO, since we don't expect a lot of calls for this.

**Review by vmaheshw — CHANGES_REQUESTED** [review]

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 1 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:58]: The field name suggests that it is rate in KB/sec, hence 1024.

Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:58]: The way I see it, the buffer for growth will be reserved by elastic task mechanism. We already have that there, right? Do we need to bring it here as well?

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 5 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:32]: I'd prefer to leave it as it is.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 3 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:48]: I was thinking about this too. What if throughput info is not available for a subset of partitions? I think it makes sense to just print a warning message in that case. Worst case scenario, the old mechanism for task count estimation will win, but a warning message would be helpful to have.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 2 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:54]: That's an interesting point. Do you think that weight should be configurable?

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 3 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:48]: There's a slight problem here though. If we want to filter the provider results and get only info for partitions with throughput > X, there won't be a way to tell if the throughput info was filtered out or it wasn't available.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 2 above)
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:54]: Yes, weight should be configurable. Please add that as well in this PR.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 3 above)
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:48]: Empty map can be treated as throughput not available. In case of Pinot, you will get success or failure. In case of Brooklin monitor, you can assume and retry.

Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:48]: Also, Throughput > X is very critical to get the results faster in case of Pinot.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 7]
```diff
@@ -27,6 +41,23 @@
...
+    allPartitions.addAll(unassignedPartitions);
+
+    // total throughput in KB/sec
+    int totalThroughput = throughputMap.entrySet().stream().
+        filter(e -> allPartitions.contains(e.getKey())).mapToInt(e -> e.getValue().getBytesInKBRate()).sum();
+    LOG.info("Total throughput in all partitions: {}KB/sec", totalThroughput);
+
+    int taskCountEstimate = (int) Math.ceil((double) totalThroughput / (TASK_CAPACITY_MBPS_DEFAULT * 1024));
+    taskCountEstimate = Math.min(allPartitions.size(), taskCountEstimate);
+    LOG.debug("Estimated number of tasks required to handle the throughput: {}", taskCountEstimate);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:60]: Make it info, as this is important for us and will not be very frequent, also add assigned and unassigned partition count

File: datastream-server/src/test/java/com/linkedin/datastream/server/TestLoadBasedTaskCountEstimator.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import com.linkedin.datastream.server.assignment.LoadBasedTaskCountEstimator;
+import com.linkedin.datastream.server.providers.FileBasedPartitionThroughputProvider;
+
+
+/**
+ * Tests for {@link LoadBasedTaskCountEstimator}
+ */
+public class TestLoadBasedTaskCountEstimator {
+  private static final String THROUGHPUT_FILE_NAME = "partitionThroughput.json";
+  FileBasedPartitionThroughputProvider _provider;
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/TestLoadBasedTaskCountEstimator.java:25]: add negative tests for throughput information not present, assigned partitions, unassigned partitions

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 9]
```diff
@@ -27,6 +41,23 @@
    */
   public int getTaskCount(ClusterThroughputInfo throughputInfo, List<String> assignedPartitions,
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:42]: One of the major issue I have with task counting using both assigned partition and unassigned partition is, sticky strategy reuses the task object and it will be tricky in this case.

This will not show up as an issue, since we will call the method with assignedPartitions as empty. Do you think, it will make more sense to expose another method `getTaskCount(throughputInfo, unassignedPartitions)` and internally call `getTaskCount(throughputInfo, Collections.emptyList(), unassignedPartitions)`. This way when we later read the code, it will not create unnecessary confusion.

**Review by vmaheshw — COMMENTED** [review]
Mostly looks good. Please add default weight change and threshold to fill (default can be 90%).

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 10]
```diff
@@ -27,6 +41,23 @@
...
       List<String> unassignedPartitions) {
-    throw new NotImplementedException();
+    Validate.notNull(throughputInfo, "null throughputInfo");
+    Validate.notNull(assignedPartitions, "null assignedPartitions");
+    Validate.notNull(unassignedPartitions, "null unassignedPartitions");
+
```
[the remark this answers is not in the document]
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:47]: This will go in `LoadBasedPartitionAssignmentStrategy`

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 9 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:42]: I think javadoc and variable naming should take care of that.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 11]
```diff
@@ -85,18 +94,19 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
...
         "Zero tasks assigned. Retry leader partition assignment.");
 
+    // Resolving cluster name from datastream group
+    String clusterName = _sourceClusterResolver.getSourceCluster(datastreamPartitions.getDatastreamGroup());
+    ClusterThroughputInfo clusterThroughputInfo = partitionThroughputInfo.get(clusterName);
+
     // TODO Get task count estimate and perform elastic task count validation
     // TODO Get task count estimate based on throughput and pick a winner
-    int maxTaskCount = 0;
+    LoadBasedTaskCountEstimator estimator = new LoadBasedTaskCountEstimator(_taskCapacityMBps, _taskCapacityUtilizationPct);
```
Comment (jzakaryan) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:103]: This were added here just to make the code compile. Final implementation of `LoadBasedPartitionAssignmentStrategy` will be published in a separate PR.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 12]
```diff
@@ -27,6 +41,23 @@
...
+    Map<String, PartitionThroughputInfo> throughputMap = throughputInfo.getPartitionInfoMap();
+
+    Set<String> allPartitions = new HashSet<>(assignedPartitions);
+    allPartitions.addAll(unassignedPartitions);
+
+    // total throughput in KB/sec
+    int totalThroughput = throughputMap.entrySet().stream().
+        filter(e -> allPartitions.contains(e.getKey())).mapToInt(e -> e.getValue().getBytesInKBRate()).sum();
+    LOG.info("Total throughput in all partitions: {}KB/sec", totalThroughput);
```
[the remark this answers is not in the document]
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java]: It's added in the previous log message.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 13]
```diff
@@ -40,8 +41,13 @@
...
+  private static final int TASK_CAPACITY_MBPS_DEFAULT = 4;
+  private static final int TASK_CAPACITY_UTILIZATION_PCT_DEFAULT = 90;
+
   private final PartitionThroughputProvider _throughputProvider;
   private final DatastreamSourceClusterResolver _sourceClusterResolver;
+  private final int _taskCapacityMBps;
+  private final int _taskCapacityUtilizationPct;
 
 
   /**
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:53]: javadocs missing.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 14]
```diff
@@ -85,18 +94,19 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
     Validate.isTrue(currentAssignment.size() > 0,
         "Zero tasks assigned. Retry leader partition assignment.");
 
+    // Resolving cluster name from datastream group
+    String clusterName = _sourceClusterResolver.getSourceCluster(datastreamPartitions.getDatastreamGroup());
+    ClusterThroughputInfo clusterThroughputInfo = partitionThroughputInfo.get(clusterName);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:99]: Just wondering, if fetchPartitionThroughputInfo object should have cluster information. Since you have the retry logic at the earlier step (`fetchPartitionThroughputInfo`, the implementation for Brooklin-monitor/Pinot will be tricky, as you will need fabric/cluster information.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 15]
```diff
@@ -5,17 +5,40 @@
...
 
 import com.linkedin.datastream.server.ClusterThroughputInfo;
+import com.linkedin.datastream.server.PartitionThroughputInfo;
 
 
 /**
  * Estimates the minimum number of tasks for a datastream based on per-partition throughput information
  */
 public class LoadBasedTaskCountEstimator {
+  private final static int BYTES_IN_KB_RATE_DEFAULT = 5;
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:25]: Make these fields configurable.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 16]
```diff
@@ -5,17 +5,40 @@
...
+   * Creates an instance of {@link LoadBasedTaskCountEstimator}
+   * @param taskCapacityMBps Task capacity in MB/sec
+   * @param taskCapacityUtilizationPct Task capacity utilization percentage
+   */
+  public LoadBasedTaskCountEstimator(int taskCapacityMBps, int taskCapacityUtilizationPct) {
+    _taskCapacityMBps = taskCapacityMBps;
+    _taskCapacityUtilizationPct = taskCapacityUtilizationPct;
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(LoadBasedTaskCountEstimator.class.getName());
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:41]: move this before the constructor.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 17]
```diff
@@ -27,6 +50,26 @@
    */
   public int getTaskCount(ClusterThroughputInfo throughputInfo, List<String> assignedPartitions,
       List<String> unassignedPartitions) {
-    throw new NotImplementedException();
+    Validate.notNull(throughputInfo, "null throughputInfo");
+    Validate.notNull(assignedPartitions, "null assignedPartitions");
+    Validate.notNull(unassignedPartitions, "null unassignedPartitions");
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:55]: One good validation will be to check if `assignedPartitions` and `unassignedPartitions` have any intersection. This is critical as we don't want two tasks to have the same partition.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 18]
```diff
@@ -27,6 +50,26 @@
...
+
+    Map<String, PartitionThroughputInfo> throughputMap = throughputInfo.getPartitionInfoMap();
+
+    Set<String> allPartitions = new HashSet<>(assignedPartitions);
+    allPartitions.addAll(unassignedPartitions);
+
+    // total throughput in KB/sec
+    int totalThroughput = allPartitions.stream().map(p ->
+        throughputMap.getOrDefault(p, new PartitionThroughputInfo(BYTES_IN_KB_RATE_DEFAULT, MESSAGES_IN_RATE_DEFAULT, p))).
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:64]: Will it make more sense to create one default PartitionThroughputInfo object, rather than N objects, since we don't it for long.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 19]
```diff
@@ -27,6 +50,26 @@
...
+    Validate.notNull(assignedPartitions, "null assignedPartitions");
+    Validate.notNull(unassignedPartitions, "null unassignedPartitions");
+
+    Map<String, PartitionThroughputInfo> throughputMap = throughputInfo.getPartitionInfoMap();
+
+    Set<String> allPartitions = new HashSet<>(assignedPartitions);
+    allPartitions.addAll(unassignedPartitions);
+
+    // total throughput in KB/sec
+    int totalThroughput = allPartitions.stream().map(p ->
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:63]: Change this for better readability. (One operator per line)

> int totalThroughput = allPartitions.stream()
        .map(p -> throughputMap.getOrDefault(p, new PartitionThroughputInfo(BYTES_IN_KB_RATE_DEFAULT, MESSAGES_IN_RATE_DEFAULT, p)))
        .mapToInt(PartitionThroughputInfo::getBytesInKBRate).sum();

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 20]
```diff
@@ -27,6 +50,26 @@
...
+        throughputMap.getOrDefault(p, new PartitionThroughputInfo(BYTES_IN_KB_RATE_DEFAULT, MESSAGES_IN_RATE_DEFAULT, p))).
+        mapToInt(PartitionThroughputInfo::getBytesInKBRate).sum();
+    LOG.info("Total throughput in all {} partitions: {}KB/sec", allPartitions.size(), totalThroughput);
+
+    double taskCapacityUtilizationCoefficient = _taskCapacityUtilizationPct / 100.0;
+    int taskCountEstimate = (int) Math.ceil((double) totalThroughput /
+        (_taskCapacityMBps * 1024 * taskCapacityUtilizationCoefficient));
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:70]: nit: if you save `_taskCapacityMBps * 1024 * taskCapacityUtilizationCoefficient` calculation, you can get rid of `_taskCapacityMBps`, `taskCapacityUtilizationCoefficient` and extra calculation.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 21]
```diff
@@ -27,6 +50,26 @@
    */
   public int getTaskCount(ClusterThroughputInfo throughputInfo, List<String> assignedPartitions,
       List<String> unassignedPartitions) {
-    throw new NotImplementedException();
+    Validate.notNull(throughputInfo, "null throughputInfo");
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:53]: Debug log to dump all the three fields for testing purpose will be very useful.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/PartitionAssignmentStrategyConfig.java [hunk 22]
```diff
@@ -54,6 +58,8 @@ public PartitionAssignmentStrategyConfig(Properties config) {
     int cfgMaxParitionsPerTask = props.getInt(CFG_MAX_PARTITION_PER_TASK, 0);
     int cfgPartitionsPerTask = props.getInt(CFG_PARTITIONS_PER_TASK, 0);
     int cfgPartitionFullnessThresholdPct = props.getIntInRange(CFG_PARTITION_FULLNESS_THRESHOLD_PCT, 0, 0, 100);
+    int cfgTaskCapacityMBps = props.getInt(CFG_TASK_CAPACITY_MBPS, 0);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/PartitionAssignmentStrategyConfig.java:61]: Just wondering, if we should get the default values to this class and get rid of optional.

The counter reasoning to this change request can be: in the future, we may have some class that does not need the default value. But, we can move to Optional at that time. Right now, even though we pass optional, the constructor of the class, sets the default value if Optional is empty.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 15 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:25]: I don't think there's need, @vmaheshw. I don't think it's something we will want to tune.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 17 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:55]: This will be done in the strategy class itself.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 21 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:53]: This will be done in the strategy class.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/PartitionAssignmentStrategyConfig.java (hunk 22 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/PartitionAssignmentStrategyConfig.java:61]: I tried to stick to the way other configs were implemented. If we want to change this, we might want to address the older ones too to keep things consistent.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 18 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:64]: You're right. We have the name for each partition if we do it in the loop, but then again, we're not using that name.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java (hunk 13 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:53]: Will be added in separate PR.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java (hunk 14 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:99]: `ClusterThroughputInfo` already has the `clusterName`. Let's take this offline, maybe I'm missing something.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 20 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:71]: Yeah but then I'll need to name that variable 😄

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java (hunk 14 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:99]: As per our discussion, this will go out in a separate PR.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java [hunk 23]
```diff
@@ -27,6 +54,30 @@
...
+    Validate.notNull(assignedPartitions, "null assignedPartitions");
+    Validate.notNull(unassignedPartitions, "null unassignedPartitions");
+    LOG.info("Assigned partitions: {}", assignedPartitions);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:60]: We need throughput info as well. There is no easy way for us to get this info otherwise.

**Review by vmaheshw — APPROVED** [review]
Other than 1 last comment, everything else LGTM.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java (hunk 23 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java:60]: @vmaheshw let me think about this. I don't want to dump this in the log file. It can be huge.

**Review by shrinandthakkar — APPROVED** [review]
